python/taichi/lang/struct.py

```diff
@@ -672,20 +672,22 @@ def __init__(self, **kwargs):
                 self.members[k] = cook_dtype(dtype)
 
     def __call__(self, *args, **kwargs):
+        """Create an instance of this struct type."""
         d = {}
         items = self.members.items()
+        # iterate over the members of this struct
         for index, pair in enumerate(items):
-            name, dtype = pair
-            if isinstance(dtype, CompoundType):
-                if index < len(args):
-                    d[name] = dtype(args[index])
-                else:
-                    d[name] = kwargs.get(name, dtype(0))
-            else:
-                if index < len(args):
-                    d[name] = args[index]
-                else:
-                    d[name] = kwargs.get(name, 0)
+            name, dtype = pair  # (member name, member type)
+            if index < len(args):  # set from args
+                data = args[index]
+            else:  # set from kwargs
+                data = kwargs.get(name, 0)
+
+            if isinstance(dtype, CompoundType) and not isinstance(
+                    data, (dict, Struct)):
+                data = dtype(data)
```

Comment:
Why do we need `dtype(data)` here? There's already `struct = self.cast(entries)` below.

Reply:
This is because when `dtype` is a struct type and `data` is a scalar, `self.cast(entries)` will raise an error.

Reply:
This is because if there is a struct member that leaves uninitialized, `self.cast` will try to cast the int `0` to it, hence raising an error. For example:

```python
import taichi as ti
ti.init(ti.cpu)

vec2 = ti.math.vec2

D = dict({"i": 3, "k":2.})
@ti.dataclass
class X:
    i: int

@ti.dataclass
class Z:
    i: int
    x: X

x = X(1)
z = Z(1)
print(z.x)
```

I have  updated the code so that the intention should be more clear.